### PR TITLE
T5518: Add basic MLD support

### PR DESCRIPTION
**Pull request by vfreex**
## Change Summary

Currently VyOS has `protocol igmp` option to enable IGMP querier and reports through FRR's pimd.

I would like to add support for IPv6 as well since FRR's IPv6 multicast functionality has significantly improved.

Enabling both MLD and IGMP on a VyOS router will allow us to turn on multicast snooping on layer-3 switches in dual-stack networks.

Example commands:

```
// Enable on interface eth0
set protocols pim6 interface eth0

// Explicitly join multicast group ff18::1234 on interface eth1
set protocols pim6 interface eth1 mld join ff18::1234

// Explicitly join source-specific multicast group ff38::5678 with source address 2001:db8::1 on interface eth1
set protocols pim6 interface eth1 mld join ff38::5678 source 2001:db8::1
```
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://vyos.dev/T5518

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: interface-definitions/protocols-mld.xml.in [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+                    <list>1 2</list>
+                  </completionHelp>
+                  <valueHelp>
+                    <format>1</format>
+                    <description>MLD version 1</description>
+                  </valueHelp>
+                 <valueHelp>
+                    <format>2</format>
+                    <description>MLD version 2</description>
+                  </valueHelp>
```
Comment (sever-sever) [inline interface-definitions/protocols-mld.xml.in]: Should there be  a default value? And which version is default?

File: src/conf_mode/protocols_mld.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
+#!/usr/bin/env python3
+#
+# Copyright (C) 2020-2023 VyOS maintainers and contributors
```
Comment (sever-sever) [inline src/conf_mode/protocols_mld.py]: Just 2023

File: src/conf_mode/protocols_mld.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        conf = config
+    else:
+        conf = Config()
+    mld_conf = {
+        'mld_conf' : False,
+        'old_ifaces'    : {},
+        'ifaces'    : {}
+    }
+    if not (conf.exists('protocols mld') or conf.exists_effective('protocols mld')):
+        return None
```
Comment (sever-sever) [inline src/conf_mode/protocols_mld.py]: The config exists and effective are old implementation of configuration. 
You should use get_config_dict ad in this example https://github.com/vyos/vyos-1x/blob/d3edda22573fb9c9d1c469f14f5a3eec9ca512a5/src/conf_mode/protocols_babel.py#L41

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: interface-definitions/protocols-mld.xml.in (hunk 1 above)
Comment (vfreex) [reply interface-definitions/protocols-mld.xml.in]: Done adding the default value.

File: src/conf_mode/protocols_mld.py (hunk 3 above)
Comment (vfreex) [reply src/conf_mode/protocols_mld.py]: done

**Review by sever-sever — APPROVED** [review]

**Comment by c-po** [comment]
Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

**Comment by vfreex** [comment]
> Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

This PR mostly follows what `protocols igmp` does. I don't know if there is a reason actually. If we believe it is better to make the daemon always run I can make a change.

File: src/conf_mode/protocols_mld.py [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+from vyos.utils.process import process_named_running
+from vyos.utils.process import call
+from vyos.template import render
+from signal import SIGTERM
+
+from vyos import airbag
+airbag.enable()
+
+# Required to use the full path to pim6d, in another case daemon will not be started
+pim6d_cmd = f'/usr/lib/frr/pim6d -d -F traditional --daemon -A 127.0.0.1'
```
Comment (c-po) [inline src/conf_mode/protocols_mld.py]: This is for IPv6, correct? Or should pim for IPv4 be used?

**Review by c-po — CHANGES_REQUESTED** [review]

**Comment by c-po** [comment]
> Any reason to not make the daemon always run as we do with bgpd, ospfd, ospf6d?

Okay, this was me thinking it'sIPv4 PIM where we can only run pimd or igmp-proxy - but this is IPv6.

Can we make it `set protocols pim6` instead? So I see no reason to not include support for this in future versions.

File: src/conf_mode/protocols_mld.py (hunk 4 above)
Comment (vfreex) [reply src/conf_mode/protocols_mld.py]: Yes, this is for IPv6. Only pim6d is needed for MLD.

**Comment by c-po** [comment]
pim6d was enabled in the base image https://github.com/vyos/vyos-build/commit/1e10c349d4cbce933e6d6373c9e011e053ec3b31

Please use frr-reload like we do in other protocols (ospf, ospfv3 etc)

**Comment by vfreex** [comment]
@c-po 

> Can we make it set protocols pim6 instead? So I see no reason to not include support for this in future versions.

Yes. This PR makes it in `set protocols mld` because we have `set protocols igmp` for IPv4 (I wanted to follow the same practice). Also this PR only adds MLD query and report support and doesn't include complete pim6 features in FRR.

> pim6d was enabled in the base image https://github.com/vyos/vyos-build/commit/1e10c349d4cbce933e6d6373c9e011e053ec3b31
> Please use frr-reload like we do in other protocols (ospf, ospfv3 etc)

Thanks. I will make a change to use frr-reload.

**Comment by vfreex** [comment]
@c-po I've updated this PR and added smoketests for it. However I don't have an environment to run smoketests. Let me know if there are other any issues.

File: data/templates/frr/daemons.frr.tmpl [hunk 5]
```diff
@@ -38,6 +39,7 @@ isisd_options="  --daemon -A 127.0.0.1
 {%- if snmp is defined and snmp.isisd is defined %} -M snmp{% endif -%}
 "
 pimd_options="  --daemon -A 127.0.0.1"
+pim6d_options=""--daemon -A ::1"
```
Comment (sever-sever) [inline data/templates/frr/daemons.frr.tmpl]: Are there three double quotes?

**Review by sever-sever — CHANGES_REQUESTED** [review]

File: data/templates/frr/daemons.frr.tmpl (hunk 5 above)
Comment (vfreex) [reply data/templates/frr/daemons.frr.tmpl]: Nice spot! Now fixed.

**Review by dmbaturin — APPROVED** [review]

File: interface-definitions/protocols-pimv6.xml.in [hunk 6]
```diff
@@ -0,0 +1,105 @@
+<?xml version="1.0"?>
+<!-- Protocol Independent Multicast for IPv6 (PIMv6) configuration -->
+<interfaceDefinition>
+  <node name="protocols">
+    <children>
+      <node name="pimv6" owner="${vyos_conf_scripts_dir}/protocols_pimv6.py">
```
Comment (c-po) [inline interface-definitions/protocols-pimv6.xml.in]: Please use only `pim6` as FRR has `pim` and `pim6` no need for an extra `v`

File: interface-definitions/protocols-pimv6.xml.in [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+                      </valueHelp>
+                      <constraint>
+                        <validator name="ipv6-address"/>
+                      </constraint>
+                    </properties>
+                    <children>
+                      <leafNode name="source">
+                        <properties>
+                          <help>Source address</help>
+                          <valueHelp>
```
Comment (c-po) [inline interface-definitions/protocols-pimv6.xml.in]: Please add a completion helper:
```
    <completionHelp>
      <script>${vyos_completion_dir}/list_local_ips.sh --ipv6</script>
    </completionHelp>
```

File: interface-definitions/protocols-pimv6.xml.in [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+                        <format>2</format>
+                        <description>MLD version 2</description>
+                      </valueHelp>
+                      <constraint>
+                        <validator name="numeric" argument="--range 1-2"/>
+                      </constraint>
+                    </properties>
+                    <defaultValue>2</defaultValue>
+                  </leafNode>
+                  <leafNode name="query-interval">
```
Comment (c-po) [inline interface-definitions/protocols-pimv6.xml.in]: Can we reduce this to `interval`? The help string is sufficient and we could try to make a common interval node.

File: interface-definitions/protocols-pimv6.xml.in [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+                      <valueHelp>
+                        <format>u32:1-65535</format>
+                        <description>Query interval in seconds</description>
+                      </valueHelp>
+                      <constraint>
+                        <validator name="numeric" argument="--range 1-65535"/>
+                      </constraint>
+                    </properties>
+                  </leafNode>
+                  <leafNode name="query-max-response-time">
```
Comment (c-po) [inline interface-definitions/protocols-pimv6.xml.in]: Can we drop `query`?

Also deci seconds is an odd interval. As sub second granularity seems to be used - can we switch to milliseconds?

File: src/conf_mode/protocols_pimv6.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    # that they will be properly rendered for the FRR config. Also this eases
+    # removal of interfaces from the running configuration.
+    interfaces_removed = node_changed(conf, base + ['interface'])
+    if interfaces_removed:
+        pimv6['interface_removed'] = list(interfaces_removed)
+
+    return pimv6
+
+
+def verify(pimv6: Optional[ConfigDict]):
```
Comment (c-po) [inline src/conf_mode/protocols_pimv6.py]: Can we please stick with the current signature and not use Optional?

**Review by c-po — CHANGES_REQUESTED** [review]

File: interface-definitions/protocols-pimv6.xml.in (hunk 6 above)
Comment (vfreex) [reply interface-definitions/protocols-pimv6.xml.in]: Done

File: interface-definitions/protocols-pimv6.xml.in (hunk 7 above)
Comment (vfreex) [reply interface-definitions/protocols-pimv6.xml.in]: Added

File: interface-definitions/protocols-pimv6.xml.in (hunk 8 above)
Comment (vfreex) [reply interface-definitions/protocols-pimv6.xml.in]: renamed

File: interface-definitions/protocols-pimv6.xml.in (hunk 9 above)
Comment (vfreex) [reply interface-definitions/protocols-pimv6.xml.in]: done. Also changed the unit to milliseconds and the value will be divided by 100 in the jinja template.

**Review by c-po — APPROVED** [review]

**Comment by vfreex** [comment]
https://github.com/vyos/vyos-documentation/pull/1077 for documentation.
https://github.com/vyos/vyos-1x/pull/2233 fixes the smoketest.

**Pull request merged**